Shopify Store, Header 의 login 모달 연결하기

xunxee opened the pull request.

xunxee committed:
Fix: eslint 오류 테스트 중

xunxee committed:
Merge branch 'develop' into feature/login

xunxee committed:
Merge branch 'develop' into feature/login

xunxee committed:
Add: Header 의 login 아이콘 클릭시 모달창이 뜨는 기본적인 액션을 추가

xunxee committed:
Add: reset css 를 추가

ChoJinmok commented on src/Header/PurchaseContainer.jsx:
```diff
@@ -18,7 +23,8 @@ export default function PurchaseContainer() {
         icon={faCircleUser}
         size="2x"
         color="red"
+        onClick={onClick}
```
저도 예전에 아이콘이나 h1, span 이런 태그들에 바로 클릭 이벤트를 넣어줬었는데 클릭이 일어나는 것은 button태그로 묶어줘서 button에 클릭 이벤트를 넣는게 좀더 semantic하게 태그를 사용하는 것 같더라구요? 이건 그냥 제 생각이라~ 한번 참고해보세용 ㅎㅎ

xunxee replied:
생각해보니 아이콘에 바로 "onClick" 을 주는 것이 semantic 하지 못한 것 같군요 ㅎㅎ
보통 클릭하면 떠올릴 수 있는 button 태그를 이용해서 감싸주어야 겠습니다! 🙏🏼

ChoJinmok commented on src/LoginPage/slice.js:
```diff
@@ -3,6 +3,7 @@ import { createSlice } from '@reduxjs/toolkit';
 const { actions, reducer } = createSlice({
   name: 'login',
   initialState: {
+    isModal: false,
```
isModal이라는 이름이 무슨 의미인지는 알 거 같지만 그래도 정확하게 써주는게 좋을 거 같아요!! isModalOpen이런식으루요? 한번 참고하세용 ㅎㅎ

xunxee replied:
isModalOpen 이 더 직관적인 것 같네요!! 감사합니당 ㅎㅎ

ChoJinmok reviewed (commented):
CSS땜에 고생이 많으시죠 😭 지금 고비들만 잘 넘기시면 무서울게 없지지않을까합니당 ㅎㅎ 팟팅!!
이번에도 slice테스트에 초기 state값이 return 되는 테스트 한번 추가해보시면 좋을 거 같습니다!

xunxee committed:
Refactor: 10월 16일의 리뷰를 반영
- 아이콘을 바로 클릭하는 것이 아니라 semantic 하게 button 태그 안에 넣어서 리팩터링
- slice 의 초기 state 값이 반환되는 테스트를 추가
- initialState 에서 isModal 을 isModalOpen 으로 바꾸어 가독성을 개선
- HeaderPage 컴포넌트에서 useCallback 을 사용하여 최적화

xunxee committed:
Modify: Header 컴포넌트의 기본적인 CSS 를 수정

xunxee committed:
Modify: SearchBarContainer 컴포넌트에 돋보기 아이콘을 추가

xunxee committed:
Add: Header 컴포넌트의 CSS를 추가
- 화면이 950px 미만부터는 줄어들지 않도록 설정
- Header 의 리스트들의 마우스 hover 효과를 추가

ChoJinmok commented on src/Header/SearchBarContainer.test.jsx:
```diff
@@ -4,10 +4,18 @@ import SearchBarContainer from './SearchBarContainer';
+  it('renders the Magnifying Glass icon', () => {
+    const { queryByTitle } = render(<SearchBarContainer />);
+
+    expect(queryByTitle('magnifyingGlass')).not.toBeNull();
```
오! 저는 FontAwesome 어떻게 테스트할까 고민했었는데 tilte로 테스트 하는 거 아이디어가 완전 좋은데요?!

xunxee replied:
이리저리 해보다가 얻어 걸렷군용 @.@
이게 그런데 맞는 방법인지는 잘 모르겟습니다 🥹

ChoJinmok commented on src/Header/SearchBarContainer.jsx:
```diff
@@ -1,8 +1,44 @@
+
+const Container = styled.div({
+  position: 'relative',
+  flexBasis: '33%',
```
flexBasis는 저도 한번도 사용안해본 속성이네요! 👍
찾아보니 width를 사용해도 가능할 거 같은데 flexBasis를 사용하신 이유가 있을까용?~

그리고 33%주신 이유는 아마 검색바를 가운데 두시려고 하신거죠?~ 그런데 이런 경우 1%가 비어서 조금 틀어지는 경우가 생길 수도 있을 거 같습니다!
calc(100% / 3)을 사용하시는게 조금 더 정확할 거 같지만 검색창 너비가 '100% / 3'을 넘어야 하는 경우에는 틀어질 수 있을 거 같기도 하네요!ㅠㅠ 한번 고민해보셔도 좋을 거 같습니다! (검색창의 너비가 넓어져도 가운데를 유지할 수 있는 안전 장치만 추가하면 좋을 듯 합니당!)

저 같은 경우에는 양쪽요소는 space-betweent으로 맞추고 가운데 요소는 absolute로 가운데 위치를 맞춰주기도 했습니당!! 정답은 없으니 한번 참고만 하세용!

ChoJinmok reviewed (commented):
CSS는 정답이 없어서~ 제 이야기는 최대한 참고만 하시고 건땅이 하신 방법에서 최대한 방법을 찾아가는게 좋을 듯 합니당 ㅎㅎ
CSS도 완전 잘하고 계신 거 같아서 걱정이 없네요 👍

xunxee committed:
Add: LoginPage 의 바깥 지점 클릭시 언마운트되는 기능을 추가

xunxee committed:
Refactor: 리뷰 내용을 반영

xunxee committed:
Refactor: login form 의 위치를 수정
- nav 바의 postion 에 fixed 를 사용하면서 login form 의 위치가 바뀌어서 이를 수정

xunxee committed:
Refactor: LoginPage 컴포넌트의 listener 함수의 로직을 수정
- 기존의 로직에서 가독성을 위하여 논리 부정 연산자를 수정

xunxee committed:
Add: LoginPage 모달창 생성시 뒷부분을 blur 처리하는 기능을 추가

ChoJinmok commented on src/Header/HeaderPage.jsx:
```diff
@@ -1,14 +1,61 @@
-    </>
+      <PurchaseContainer
+        onClick={handleToggle}
+        refUserIcon={refUserIcon}
```
forwardRef에 대해 한번 찾아보셔도 좋을 거 같아요! ref를 props로 전달해주는 방법인데 건땅이 하신 거랑 한번 비교해보시면 좋을 거 같아요!!
그리고 useImperativeHandle도 한번 찍먹해보시길 바랍니당 ㅎㅎ(진짜 찍먹만 하기!)

xunxee replied:
HTML 엘리먼트가 아니라 React 컴포넌트에서 ref prop을 사용할 때 forwardRef() 를 사용할 수 있군용!

ChoJinmok replied:
음~ 내일 한번 같이 이야기해봐도 좋을 거 같네요 ㅎㅎ

ChoJinmok commented on src/Header/PurchaseContainer.jsx:
```diff
@@ -1,24 +1,46 @@
+        <FontAwesomeIcon
+          title="circleUser"
+          icon={faCircleUser}
+          size="2x"
```
FontAwesome의 아이콘은 font-size로도 크기가 조절 가능합니당!~ 혹시나 더 세세하게 조절할 필요가 있다면 참고하세용!

ChoJinmok commented on src/LoginPage/LoginPage.jsx:
```diff
@@ -1,20 +1,55 @@
+  const refLogin = useRef();
+
+  function listener({ target }) {
+    if (refLogin.current.contains(target)) { return; }
```
```suggestion
    if (refLogin.current.contains(target)) return; 
```

한줄에 다 작성할 때는 이렇게 작성해주셔도 될겁니당!!ㅎㅎ

xunxee replied:
if 문의 가독성을 위해서 중괄호를 생략하지 않고 항상 사용했는데 한 줄로 할 때는 생략하는 편이 좋을 것 같군용 ㅎㅎ

ChoJinmok replied:
그런데 다희님이 그렇다면 그런거라.. ㅎ

ChoJinmok commented on src/Footer/FooterPage.jsx:
```diff
@@ -1,10 +1,16 @@
 import LogoContainer from './LogoContainer';
 
+const Container = styled.div({
+  zIndex: '2',
```
zIndex가 없으면 제대로 동작을 하지 않나용?~🤔

xunxee replied:
아앗 이거 제가 blur 처리하면서 코드를 수정하고 해보다가 삭제를 하지 못햇네요 🥹

ChoJinmok reviewed (commented):
크게 리뷰드릴게 없네요!👍 TDD! 가보시죵!~

xunxee committed:
Add: "Log In" 버튼과 "Sign Up" 버튼 클릭시 dispatch가 일어나는 액션을 테스트 및 구현

xunxee committed:
Add: login에서 “Log In” 버튼 클릭시 일어나는 submit 이벤트를 추가중
  - LoginFormContainer 컴포넌트의 handleSubmit()를 useCallback을 사용하여 리팩터링
  - slice.js에서 requestLogin 함수를 추가
  - requestLogin  함수 안에서 사용되는 postLogin 함수를 api.js 에서 구현중

ChoJinmok commented on src/LoginPage/LoginForm.jsx:
```diff
@@ -47,7 +48,10 @@ export default memo(({
         onChange={handleChange}
       />
-      <button type="submit">
+      <button
+        type="button"
+        onClick={onSubmit}
+      >
```
type을 button으로 했을 때와 submit으로 했을 때 차이를 한번 고민해보시면 좋을 거 같습니당!!
type을 submit으로 하고 form에 onSubmit이벤트를 줘서 form의 onSubmit이벤트에 onSubmit함수를 넣어주시는 게 좋을 거 같네요!~ㅎㅎ
아래쪽은 type을 button으로 두는 게 맞는지 아닌지도 한번 고민해보시면 좋을 거 같습니다 ㅎㅎ

xunxee replied:
form 안에 존재하는 button 타입을 submit 으로 설정하고 button을 클릭하면 form에 설정한 onSubmit 이벤트가 실행되도록 하는 것이 맞겠군요 ㅎㅎ
button의 type을 button으로 설정하면 button을 클릭해도 form의 이벤트는 실행되지 않네요

재로드를 방지하는 것으로는 button을 눌렀을 때 발생하는 onSumbit 이벤트로 호출되는 함수에 preventDefault를 추가하여 방지해야겠습니당

ChoJinmok commented on src/LoginPage/LoginFormContainer.jsx:
```diff
@@ -31,13 +33,19 @@ export default function LoginFormContainer() {
   }, [dispatch]);
 
+  const handleSubmit = useCallback(() => (
+    isLogin
+      ? dispatch(requestLogin()) : dispatch(requestSignup())
```
```suggestion
    isLogin
      ? dispatch(requestLogin())
      : dispatch(requestSignup());
```

이런식으로 작성하면 조금 더 가독성이 좋을 거 같습니당 ㅎㅎ (취향입니다..ㅎ)
그런데 if문을 사용하지 않고 삼항조건연산자를 사용하는 장점이 있을까용??🤔
저도 뭐가 더 좋을 지 모르겠어서 같이 고민해보면 좋을 거 같네요!

```suggestion
    if (isLogin) dispatch(requestLogin());
    dispatch(requestSignup());
```

xunxee replied:
삼항 조건 연산자 표현식이 if문과 다른 차이점은, 삼항 조건 연산자 표현식은 값처럼 사용할 수 있다는 점인데
그래서 어떠한 조건에 따른 값을 결정하는 부분이라 삼항 조건 연산자를 사용해보았네용
그런데 삼항 조건 연산자가 기본적으로 가독성이 좋지 못해서 if문의 사용도 좋을 것  같네요!

ChoJinmok commented on src/LoginPage/slice.js:
```diff
@@ -34,8 +43,24 @@ const { actions, reducer } = createSlice({
   changeLoginFields,
 } = actions;
 
+export function requestLogin() {
```
requestLogin도 테스트코드 작성해보세용!~ㅎㅎ

xunxee replied:
넵 테스트코드 한 번 진행하겠습니다 ㅎㅎ

ChoJinmok reviewed (commented):
api관련해서는 직접 이야기 해보는게 좋을 거 같아서 따라 리뷰는 달지 않았습니당!~ 같이 한번 이야기해보시죵!ㅎㅎ

xunxee committed:
Add: 서버 통신을 위한 fireBase의 설정을 추가
- .env를 사용하여 API_KEY를 노출되지 않도록 설정
- webpackconfig 파일에 plugins를 사용하여 process.env로 환경변수를 설정

xunxee committed:
Refactor: 10월 28일의 리뷰를 반영
- LoginForm 컴포넌트와 LoginFormContainer의 구조를 수정
- 컴포넌트 구조 수정에 따른 테스트 코드를 수정

xunxee committed:
Add: api에 postRefreshToken 함수를 추가 및 에러 처리 작업 진행중
- api.js의 postLogin 함수에서 postRefreshToken 함수를 실행하는 로직을 추가
- slice.js에서 에러 처리를 위한 로직을 진행중

ChoJinmok commented on src/services/api.js:
```diff
@@ -1,16 +1,52 @@
+
+  const { refreshToken } = data;
+
+  const { uid } = postRefreshToken(refreshToken);
```
여기도 await가 필요할 거 같네용?!🤔

xunxee replied:
엇 네넹 수정하겠습니당! 감사합니다

ChoJinmok reviewed (commented):
api.js도 테스트 코드 한번 작성해보시죵!!ㅎㅎ

xunxee committed:
Add: api의 테스트를 위한 fixtures 폴더를 추가
- fixtures 폴더와 refreshToken.js 파일을 생성

xunxee committed:
Add: slice 테스트에 setRefreshToken의 액션을 테스트 및 구현

xunxee committed:
Add: storage 파일의 saveItem 함수의 액션을 테스트 및 구현

xunxee committed:
Add: slice의 setAccountInfo 함수의 액션을 테스트 및 구현

ChoJinmok commented on .gitignore:
```diff
@@ -15,3 +15,6 @@ node_modules
 coverage
+
+## EnviromentPlugin
+.env
```
push하고 Files changed 확인하는 습관 들이시면 좋습니당 ㅎㅎ

xunxee replied:
앗 에러가 발생했겠군요 ㅠㅠ
Files changed 확인을 한 번씩 꼭 해야겠습니다 ㅠㅠ

ChoJinmok commented on src/LoginPage/LoginPage.jsx:
```diff
@@ -1,20 +1,53 @@
   width: '418px',
   height: '420px',
-  backgroundColor: '#000000',
+  margin: '0',
```
margin에 0을 주신 이유가 있을까요?~🤔

xunxee replied:
아고 이 부분은 전에 blur 처리 구현하면서 여러가지 넣어보았던 부분 인 것 같네요 ㅠㅠ
삭제해야겠습니다!!

ChoJinmok reviewed (commented):
쉽지 않으실텐데 ㅠㅠ 팟팅입니다!!

xunxee committed:
Add: CI와 codeceptJS 사용을 위한 환경설정

xunxee committed:
Add: api.js, postLogin 함수의 예외 처리를 위한 액션을 테스트 및 구현
- fixtures에 loginFields.js와 loginFailData.js 파일을 추가
- postLogin 함수 내부의 예외 처리를 확인하기 위해 rejects.toThrowError를 사용

xunxee committed:
Refactor: jsconfig.json의 No newline at end of file 에러를 해결

ChoJinmok commented on src/LoginPage/slice.js:
```diff
@@ -30,12 +48,68 @@ const { actions, reducer } = createSlice({
+    } = getState();
+
+    try {
+      const data = await postLogin({ email, password });
```
```suggestion
const { refreshToken, localId: uid } = await postLogin({ email, password });
```
저는 요렇게 한방에 적어주는게 조금더 직관적이기는 하더라구용? 참고하세용!

xunxee replied:
엇 그렇네요! data를 변수 선언해서 받는 과정을 넘어갈 수 있겠군요 ㅎㅎ
감사합니당🙏🏼

ChoJinmok commented on .github/workflows/ci.yml:
```diff
@@ -0,0 +1,26 @@
+
+    strategy:
+      matrix:
+        node-version: [12.x]
```
github action 에러 코드를 보니까 npm ci에서 에러가 나더라구요?? 그래서 찾아보니까 node-version을 14로 바꿔줘야한다는 것 같습니당!!
```suggestion
        node-version: [14.x]
```
이렇게 한번 바꿔주시죵!

xunxee replied:
덕분에 검색을 해보니,
node12는 22년 4월부터 지원이 중단되어서 노드 12 대신 16으로 업데이트를 하라는 부분들이 있던데
최신 버전은 불안정한 부분들이 있어서 14버전으로 바꾸어주는 걸까요?

xunxee committed:
Refactor: 2022 11 06의 리뷰를 반영

- slice.js, requestLogin 함수의 로직을 개선
- github/ workflows/ci.yml의 node-version을 12에서 14로 변경

xunxee committed:
Add: LoginPage의 addEventLister의 액션을 테스트 및 구현

xunxee committed:
Refactor: LoginForm 컴포넌트의 에러 메시지 렌더링을 테스트 및 구현

ChoJinmok reviewed (commented):
오 eventListener 이렇게 테스트할 수 있겠군요?! 아주 멋집니당 👍
조금더 생각해보셨으면 좋겠는 점은 'events에 let을 사용안하고 할 수 있는 방법이 있을까?'입니당 (api test에서 fetch를 mocking해준 방법 참고)
또 한가지는 spyOn을 사용해서 하는 방법도 한번 생각해보셨으면 좋겠네용 ㅎㅎ (stoarage test 참고)

xunxee committed:
Refactor: 2022 11 08의 리뷰를 반영
- 기존 let events 객체를 global을 이용하여 let을 사용하지 않고 코드를 구현할 수 있도록 리팩터링

xunxee committed:
Refactor: storage.js 함수의 액션을 테스트하는 코드를 리팩터링
- saveItem 함수의 액션을 테스트하는 expect를 추가

xunxee committed:
Refactor: slice.js, requestLogin 함수의 try와 catch 상황에서의 액션을 테스트
- ‘when login  is successful’과 when login fails’의 상황으로 나누어 코드를 테스트

xunxee committed:
Refactor: api.js의 사용하지 않는 함수를 정리

xunxee committed:
Refactor: HeaderPage 컴포넌트의 액션을 리팩터링
- isModalOpen의 value에 따라 분기되는 테스트를 추가

ChoJinmok commented on src/services/storage.test.js:
```diff
@@ -1,15 +1,33 @@
 describe('storage', () => {
-  jest.spyOn(localStorage, 'setItem');
+  jest.spyOn(
+    Object.getPrototypeOf(localStorage),
```
오.. Object.getProtoTypOf로도 프로토타입에 접근할 수 있군요?!👍

xunxee replied:
https://url.kr/4kp3ej
에러 때문에 getPrototypeOf를 사용했는데 한 번 보셔도 좋을 것 같네용

ChoJinmok commented on src/Header/HeaderPage.test.jsx:
```diff
@@ -12,14 +20,57 @@ describe('HeaderPage', () => {
+    expect(dispatch).toBeCalledWith(setIsModalOpen());
+  });
+
+  context('when isModalOpen is true', () => {
```
저도 다희님께 리뷰받았던 지점인데 context를 사용한다는 건 분기로 나뉘는 거라서 한쪽에만 작성해주면 안된다고 하더라구요??
그래서 위에서도 '모달창이 닫혔을 때'로 context를 따로 작성해주시는게 좋을 거 같네용 ㅎㅎ
state로 상황을 적어줘도 괜찮긴 하지만 'when modal is open'이런식으로 상황을 적어주는게 조금 더 적절해보입니당 ㅎㅎ 참고하세용!!

xunxee replied:
그렇군요!
context를 'when modal is open'이랑 'when modal is colsed'로해서 추가해야겠습니당
감사합니닷 🐥

ChoJinmok commented on src/LoginPage/LoginPage.test.jsx:
```diff
@@ -27,19 +27,19 @@ describe('LoginPage', () => {
       document.removeEventListener = jest.fn((event) => {
-        delete events[event];
+        delete global.events[event];
       });
```
```javascript
const mockEventListener  = (events) => {
  document.addEventListener = jest.fn((event, callback) => {
    events[event] = callback;
  }

  document.removeEventListener = jest.fn((event) => {
    delete events[event];
  }
}

// ...

context('when click the modal', () => {
  beforeEach(() => {
    const events = {};

    mockEventListener(events);
  }

  // ...
}
```

ChoJinmok replied:
```javascript
jest.spyOn(EventTarget.prototype, 'addEventListener');

const mockEventListener  = (events) => {
  EventTarget.prototype.addEventListener = jest.fn((event, callback) => {
    events[event] = callback;
  }

  EventTarget.prototype.removeEventListener = jest.fn((event) => {
    delete events[event];
  }
}

// ...

context('when click the modal', () => {
  beforeEach(() => {
    const events = {};

    mockEventListener(events);
  }

  // ...
}
```

ChoJinmok replied:
머리속으로만 생각한거라 잘 동잘할지는 모르겠지만 내일 같이 이야기해보면 좋을 듯 합니당 ㅎㅎ

xunxee replied:
"Assignment to property of function parameter" 에러가 발생하더라고요
함수의 매개변수에 원시값을 주는 것은 상관없지만, 매개변수에 객체를 전달하고 값을 재할당할 때  원본이 변경되다보니(위의 경우에는 원본값을 바꾸기 위함이지만) eslint에서는 이 sideEffect를 잡아 내는군용

ChoJinmok replied:
```javascript
jest.spyOn(EventTarget.prototype, 'addEventListener');

const mockEvents = () => {
  const events = {};

  EventTarget.prototype.addEventListener = jest.fn(
    (event, callback) => {
      events[event] = callback;
    }
  );

  EventTarget.prototype.removeEventListener = jest.fn(
    (event) => {
      delete events[event];
    }
  );

  return events;
}

// ...

context('when click the modal', () => {
  it((' ... ') => {
    const events = mockEvents();

    // ...

  });
});
```

ChoJinmok commented on src/LoginPage/LoginPage.test.jsx:
```diff
@@ -23,4 +23,56 @@ describe('LoginPage', () => {
+      });
+    });
+
+    context('when click the modal', () => {
```
여기도 아래와 비슷하게 'when click inside the modal'로 작성하시는게 더 좋을 거 같아요!ㅎㅎ

xunxee replied:
제가 보아도 그렇게 바꾸어주는 것이 더 직관적으로 확인할 수 있는 것 같네요🙏🏼
감사합니당

ChoJinmok commented on src/LoginPage/LoginPage.test.jsx:
```diff
@@ -23,4 +23,56 @@ describe('LoginPage', () => {
+          />
+        ));
+
+        const $outside = document.getElementsByClassName(
+          'outsideTheModal',
+        );
```
코드를 보니까 LoginPage.jsx에 DeleteAll이라는 div가 있더라구요! 이 요소를 가지고 와서 클릭해도 테스트코드가 잘 잘동할 거 같습니당!

xunxee replied:
엇 그러면 LoginPage.jsx에서 render로 getByTestId 사용하여서 테스트할 수 있겠습니당!
test 하는데 getElementByClassName을 사용하는 것 보다 render하여 반환된 메서드를 바로 사용해야겠군용

ChoJinmok reviewed (commented):
CI 초록체크 너무 좋네요!!ㅎㅎ

xunxee committed:
Refactor: LoginPage의 addEventListener 테스트를 리팩터링
- mockEvent 함수를 만들어서 기존의 global.events를 const 키워드의 events 객체로 변경
- 기본의 코드에서 함수형으로 바꾸어 가독성을 개선
- 기존의 getElementById 메서드를 사용한 부분에서 LoginPage의 DeleteAll div를 선택하는 것으로 변경

xunxee committed:
Add: slice.js, setIsLogoutModalOpen 함수의 동작을 테스트 및 구현
- slice.js, initialState의 isModalOpen state의 이름을 → isLoginModalOpen으로 변경
- slice.js, setIsModal 함수의 이름을 → setIsLoginModal로 변경

xunxee committed:
Add: LoginPage와 LogoutPage의 기본적인 액션을 테스트 및 구현
- HeaderPage.jsx에 기본적인 LogoutPage 함수를 구현
- refreshToken과 isLoginModalOpen 상태 값을 사용한 조건부 렌더링

xunxee committed:
Add: LogoutPage의 기본적인 렌더링을 테스트 및 구현

xunxee committed:
Add: LogoutPage.jsx의 addEventListener의 액션을 테스트 및 구현

xunxee committed:
Add: LoginPage.jsx의 x버튼 클릭시 액션을 추가
- 로그인 페이지의 닫힘 버튼 클릭시 modal 창이 닫히는 액션을 추가

xunxee committed:
Add: LogoutPage.jsx의 x버튼 클릭시 액션을 추가
- 로그아웃 페이지의 닫힘 버튼 클릭시 modal 창이 닫히는 액션을 추가

xunxee committed:
Add: slice.js의 logout 함수의 액션을 테스트 및 구현

xunxee committed:
Add: LogoutPage.jsx, Log out 버튼 클릭시 액션을 테스트 및 구현
- Log out 버튼 클릭시, refreshToken을 초기화하는 액션을 테스트

xunxee committed:
Add: slice.js, ClearLoginFields 함수의 액션을 테스트 및 구현
- initialLoginFields 변수를 생성하여 loginFields의 초깃값으로 사용

xunxee committed:
Add: 로그인시 loading 처리를 추가
- slice.js, requestLogin 함수의 액션에서 changeLoginFields 함수를 추가하여 loading 처리
- slice.js, requestLogin 함수의 로직의 추가로 테스트 코드를 수정

xunxee committed:
Add: api.js, postSignUp 함수의 액션을 테스트 및 구현
- postSignUp 함수가 사용될 LoginForm 컴포넌트의 input 속성을 변경
  - 기존 placeholder에 한글 안내를 추가(성, 이름)
- api.test에서 사용한 mock/api.js에 postSignUp 함수를 추가
- api.test, postSignUp 함수에서 사용할 fixtures(signUpFailData.js)를 추가

ChoJinmok commented on src/Header/HeaderPage.jsx:
```diff
@@ -31,22 +34,32 @@ const Container = styled.div({
+        <LogoutPage
+          onClick={handleToggle}
+        />
+      )}
```
```javascript
{
  isLoginModalOpen && (
    refreshToken
      ? <LoginPage onClick={handleToggle} />
      : <LogoutPage onClick={handleToggle} /> 
  )
}
```

ChoJinmok replied:
```javascript
const SignPage = refreshToken ? LoginPage : LogoutPage;

// ...

{isLoginModalOpen && <SignPage onClick={handleToggle} />}
```

ChoJinmok replied:
줄인다고 줄여봤는데도 단축평가에 삼항조건연산자는 좀 복잡해보이는 거 같습니당 ㅠㅠ
그래서 아래 방법으로 return문 위에서 SignPage 정의한다음 아래에는 간단하게 작성하는게 조금 더 깔끔해보이네용

ChoJinmok commented on src/Header/HeaderPage.test.jsx:
```diff
@@ -46,31 +46,60 @@ describe('HeaderPage', () => {
-        },
-      }));
+        }));
+      });
```
비슷한 beforeEach가 계속 반복되니까 given 사용하셔서 리팩터링 하시면 좋을 거 같습니당 ㅎㅎ

xunxee replied:
저는 함수형으로 만들어서 리팩터링을 하려고 했는데,
given으로도 한 번 해봐도 좋을 것 같네요!
감사합니다 ㅎㅎ

ChoJinmok commented on src/LoginPage/slice.js:
```diff
@@ -4,29 +4,31 @@ import { postLogin } from '../services/api';
   name: 'login',
   initialState: {
-    isModalOpen: false,
+    isLoginModalOpen: false,
```
이제 로그아웃까지 함께 표현을 할 수 있는 단어가 좋을 듯 합니당!!
isAccountModalOpen, isSignModalOpen 이런 단어가 어떠신가용?!

xunxee replied:
slice가 나눠져 있다보니, isModalOpen으로 다시 상태를 관리할까 했는데
나중에 다른 컴포넌트에서 혹시 modal 창을 만들게 될 수도 있으니깐
로그인, 로그아웃을 표현하는 변수명이 나을 것 같군요!
감사합니당👍

ChoJinmok reviewed (commented):
Logout Page가 Login Page랑 너무 비슷해서 아예 컴포넌트를 나눠주는 것보다 하나의 컴포넌트로 처리해주는건 어떻게 생각하시나용?~ 같이 이야기해봐도 좋을 거 같습니당 ㅎㅎ

xunxee committed:
Add: Vercel 사용을 위한 package.json 설정 추가

xunxee committed:
Refactor:  LoginPage.jsx에 LoginPage 모달창과 LogoutPage 모달창을 같이 관리하도록 리팩터링
- data-testid의 이름에 따라 css 스타일링을 다르게 하도록 Container를 구성

xunxee committed:
Refactor: LoginPage.jsx의 DeleteAll을 수정
- LoginPgae와 LogoutPage에 따라 blur 효과를 선택하여 적용

xunxee committed:
Refactor: HeaderPage.test.jsx의 beforeEach 코드를 개선
- given을 사용해서 refreshToken을 관리

xunxee committed:
Refactor: slice.js의 initialState의 값과 reducers의 변수명을 수정
- isLoginModalOpen을 isAccoutModalOpen으로 변경
- setIsLoginModalOpen을 setIsAccountModalOpen으로 변경

xunxee committed:
Add: slice.js, initialState의 구조를 변경
- slice.js, initialState의 logFields의 구조를 변경
- changeLoginFiedls 함수의 로직을 변경된 loginFields에 맞도록 리팩터링
- requestLogin 함수의 로직을 변경된 loginFields에 맞도록 리팩터링

ChoJinmok commented on src/LoginPage/LoginPage.jsx:
```diff
@@ -13,23 +13,38 @@ const DeleteAll = styled.div({
-  color: '#EAEAEA',
-  backgroundColor: '#000000',
-});
+const Container = styled.div(
+  {
+    position: 'absolute',
+    color: '#EAEAEA',
+    backgroundColor: '#000000',
+  },
+  ({ 'data-testid': name }) => (
+    name === 'LoginPage' ? ({
+      top: '50%',
+      left: '50%',
+      transform: 'translate(-50%, -50%)',
+      width: '418px',
+      height: '420px',
+    }) : ({
+      top: '74px',
+      right: 0,
+      width: '150px',
+      height: '100px',
+    })
+  ),
+);
```
```javascript
const Container = styled.div(({ 'data-testid': name }) => ({
  position: 'absolute',
  color: '#EAEAEA',
  backgroundColor: '#000000',
  ...(
    name === 'loginPage'
      ? {
        top: '50%',
        left: '50%',
        transform: 'translate(-50%, -50%)',
        width: '418px',
        height: '420px',
      } : {
        top: '74px',
        right: 0,
        width: '150px',
        height: '100px',
      }
  ),
}));
```

ChoJinmok replied:
```javascript
const Container = styled.div(({ 'data-testid': name }) => {
  const loginPageLoaction = {
    top: '50%',
    left: '50%',
    transform: 'translate(-50%, -50%)',
    width: '418px',
    height: '420px',
  };

  const logoutPageLacation = {
    top: '74px',
    right: 0,
    width: '150px',
    height: '100px',
  };

  return {
    position: 'absolute',
    color: '#EAEAEA',
    backgroundColor: '#000000',
    ...(name === 'loginPage' ? loginPageLoaction : logoutPageLacation),
  };
});
```

ChoJinmok replied:
보통 이런식으로 함수형 컴포넌트 만들듯이 만들어주시면 작동할겁니당!!ㅎㅎ 참고하세요!

xunxee replied:
오... styled.div 이것도 함수 처럼 사용하면 되는군요🥹
emotio 공식문서에 'Object styles'을 보면서 작성했었는데
위와 같은 방법이 훨씬 아름답네요! 감사합니당

ChoJinmok commented on src/LoginPage/LoginPage.jsx:
```diff
@@ -4,14 +4,21 @@ import { useRef, useEffect } from 'react';
+    name === 'outsideTheModal'
+      ? { backdropFilter: 'blur(0.8px)' }
+      : { backdropFilter: 'blur(0px)' }
+  ),
```
이 부분도 리팩터링 한번 해보시면 좋을 거 같습니당!

ChoJinmok commented on src/LoginPage/LoginPage.jsx:
```diff
@@ -41,20 +63,46 @@ export default function LoginPage({ onClick }) {
+      {!refreshToken && (
+        <DeleteAll data-testid="outsideTheModal">
+          <Container
+            data-testid="LoginPage"
+            ref={refLogin}
+          >
+            <button
+              type="button"
+              onClick={onClickToggle}
+            >
+              X
+            </button>
+            <div>Logo</div>
+            <LoginFormContainer />
+          </Container>
+        </DeleteAll>
+      )}
+      {refreshToken && (
+        <DeleteAll>
+          <Container
+            data-testid="LogoutPage"
+            ref={refLogin}
+          >
+            <button
+              type="button"
+              onClick={onClickToggle}
+            >
+              X
+            </button>
+            <div>LogoutPage</div>
+            <button
+              type="button"
+              onClick={onClickLogout}
+            >
+              Log out
+            </button>
+          </Container>
+        </DeleteAll>
+      )}
+    </>
```
```jsx
<DeleteAll data-testid={refreshToken ? null : "outsideTheModal"}>
  <Container
    data-testid={refreshToken ? "LogoutPage : "LoginPage"}
    ref={refLogin}
  >
    <button
      type="button"
      onClick={onClickToggle}
    >
      X
    </button>
    <div>{refreshToken ? 'LogoutPage' : 'Logo'}</div>
    {
    refreshToken 
      ? (
      <button
        type="button"
        onClick={onClickLogout}
      >
        Log out
      </button> 
      )
       : <LoginFormContainer />
     }
  </Container>
</DeleteAll>
```

ChoJinmok replied:
코드를 반복적으로 적는 것보다는 이런식으로 작성해보시는 건 어떤가용?~

xunxee replied:
이렇게 삼항 조건 연산자를 사용해서 리팩터링을 하니
가독성이 정말 좋아지네요! 감사합니당🙏🏼
다음에도 비슷해지는 컴포넌트의 관리를 이렇게 해야할 것 같아요!

ChoJinmok commented on src/LoginPage/slice.js:
```diff
@@ -5,17 +5,31 @@ import { postLogin } from '../services/api';
+    value: '',
+    checkMessage: '',
+  },
+  error: {
+    value: '',
+  },
```
에러는 string이 아니라 객체로 작성하신 이유가 있을까용?~🤔

xunxee replied:
같이 이야기했듯이 error의 value를 담을 때
slice.js의 changeLoingFields 함수를 통해서 상태 값을
바꾸다보니, 바뀐 loginFields의 형식에 맞춰서 변경하게 되었었네요.
이제는 error의 value 만을 담당하는 reducers 함수를 만들어 보아야겠습니다! 🐥

ChoJinmok commented on src/LoginPage/slice.js:
```diff
@@ -5,17 +5,31 @@ import { postLogin } from '../services/api';
-  error: '',
+  email: {
+    value: '',
+    checkMessage: '',
```
invalidCheckMessage라고 더 구체적으로 적어주시는 게 좀 더 명확해보입니당 ㅎㅎ

xunxee replied:
그러네요, invalid를 붙이는 것이 맞는 것 같습니다 ㅎㅎ 감사합니당
정말 변수명 짓는게 항상 어렵군용

xunxee committed:
Refactor: LoginPage.jsx, styled 함수를 리팩터링
- DeleteAll과 Container 함수의 코드를 리팩터링

xunxee committed:
Refactor: LoginPage.jsx, LoginPage 함수의 반환 값을 수정
- 기존의 LoginPage와 LogoutPage를 삼항 조건 연산자를 사용해 코드를 리팩터링

xunxee committed:
Add: slice.js, changeLoginErrorMessage 함수의 액션을 테스트 및 구현
- slice.js, requestLogin 함수에서 error 메시지를 바꿀 함수로 사용될 changeLoginErrorMessage 함수를 구현

xunxee committed:
Refactor: slice.js, requestLoing 함수의 로직을 리팩터링
- requestLogin 함수에서 loginFields의 error를 바꾸기 위해 사용하던 함수를 changeLoginErrorMessage 함수로 변경

xunxee committed:
Refactor: LoginFormContainer 컴포넌트에서 handleClickToggle 함수의 액션을 변경
- 로그인 페이지와 회원가입 페이지를 전환할 때 기존 input 값을 초기화하는 clearLoginFields 함수를 추가

xunxee committed:
Refactor: slice.js, initialState의 변수명을 수정
- loginFields의 checkMessage 변수명을 invalidCheckMessage로 변경

xunxee committed:
Add: LoginFormContainer 컴포넌트의 handleSigUpValid의 기본적인 액션을 테스트 및 구현

ChoJinmok commented on src/LoginPage/LoginForm.jsx:
```diff
@@ -35,6 +36,8 @@ export default memo(({
               placeholder="성(Last Name)"
               value={lastNameValue}
               onChange={handleChange}
+              // onBlur={onBlur}
```
이 부분은 필요 없는 주석같군용!?🤔

ChoJinmok commented on src/LoginPage/LoginPage.jsx:
```diff
@@ -4,43 +4,52 @@ import { useRef, useEffect } from 'react';
-const DeleteAll = styled.div(
-  {
+const DeleteAll = styled.div(({ 'data-testid': name }) => {
+  const LoginPageLocation = {
+    backdropFilter: 'blur(0.8px)',
+  };
+
+  const LogoutPageLocation = {
+    backdropFilter: 'blur(0px)',
+  };
```
```javascript
{

// ...

...(name === 'outsideTheModal' && { backdropFilter:  'blur(0.8px)' })
}
```

ChoJinmok replied:
이렇게 작성하실수도 있을 거 같아용!ㅎㅎ 참고하세용!

xunxee replied:
알려주신 방법이 좋은 방향인 것 같습니당!
감사합니다👍
논리곱, 논리합!

ChoJinmok commented on src/LoginPage/LoginFormContainer.jsx:
```diff
@@ -27,12 +28,17 @@ export default function LoginFormContainer() {
     dispatch(changeLoginFields({ name, value }));
   }, [dispatch]);
 
+  const handleSignUpValid = useCallback(() => {
```
`handleSignUpValidCheck`나 `handleCheckSignUpValid` 함수명처럼 동사가 들어가면 좀 더 좋을 거 같습니당 ㅎㅎ

xunxee replied:
앗 그러네요! 감사합니당 ㅎㅎ

xunxee committed:
Add: slice.js, checkSignUpValid 함수의 기본적인 액션을 테스트 및 구현
- 회원가입 모달에서 onBlur 이벤트가 발생시 LoginFormContainer.jsx의 handleSignUpValid 함수가 호출되며 이때 checkSignUpValid 함수가 실행

xunxee committed:
Refactor: LoginPage.jsx의 DeleteAll 함수의 로직을 수정
- 기존의 backdropFilter 주는 방법을 논리곱 연산자를 사용하여 리팩터링

xunxee committed:
Refactor: LoginFormContainer 컴포넌트의 handleSignUpValid 함수의 변수명을 수정
- handleSignUpValid 함수명을 동사를 추가하여 handleCheckSignUpValid로 변경

xunxee committed:
Refactor: fixtures의 파일 이름과 변수명을 수정
- loginFields.js를 postLoginData.js로 수정

xunxee committed:
Add: fixtures 디렉터리에 initialLoginFields.js를 생성
- initialLoginFields는 loginFields의 구조를 가지고 있는 파일로 테스트 코드에서 이를 사용하기 위해 모듈화를 진행

xunxee committed:
Add: LoginForm 컴포넌트의 input에 값이 없는 상태로 onBlur 이벤트가 발생시 액션을 테스트 및 구현
- input에 값이 없을 경우 LoginFormContainer 컴포넌트의 handleCeckSignUpValid 함수가 실행
- checkSignUpValid 함수의 결과로 함수를 반환하고, validationCheckList 함수의 로직을 통해서 최종적으로 invalidCheckMessage의 value를 결정하게 된다.

xunxee committed:
Add: LoginForm 컴포넌트의 나머지 input에도 onBlur 이벤트가 발생시 기본적인 액션을 테스트 및 구현
- invalidCheckMessage가 있을 때에만 화면에 랜더링되도록 논리곱 연산자를 사용하여 구현
- 로그인 모달에서는 email과 password input의 onBlur 이벤트가 발생하지 않도록 삼항 연산자를 사용하여 이벤트를 처리

xunxee committed:
Refactor: Loginform의 invalidCheckMessage 테스트를 forEach 메서드를 사용하여 수정
- LoginForm.test.jex의 기존의 invalidCheckMessage 테스트를 forEach를 사용하여 반복되는 테스트 코드를 가독성있게 수정 작업

xunxee committed:
Add: sclie.test.js에 makeLoginFields 함수를 추가
- sclie.test.js에서 loginFields의 반복적인 코드 작성을 개선하기 위해 loginFields를 반환하는 함수를 작성

xunxee committed:
Refactor: slice.js, checkSignUpValid 함수의 로직을 수정
- 기존 checkSignUpValid 함수의 인자로 value를 추가하여, 함수를 반환하고 인자로 받던 getState를 인자로 받지 않도록 로직을 수정
- checkSignUpValid 함수의 인자를 RORO 패턴으로 수정

xunxee committed:
Add: slice, checkSignUpValid 함수의 이메일의 유효성을 검사하는 액션을 테스트 및 구현
- 정규표현식을 사용하여 회원가입시 이메일의 유효성을 검사

ChoJinmok commented on src/LoginPage/LoginForm.jsx:
```diff
@@ -25,7 +37,7 @@ export default memo(({
 
   function handleSignUpValid({ target: { name } }) {
-    onBlur(name);
+    onBlur({ name });
```
매개변수가 하나일 때는 꼭 RORO패턴을 고집할 필요는 없을수도 있을 거 같아요!~ 선택사항이긴해서 건땅만의 기준을 한번 잡아보시는 것도 좋을 거 같습니당 ㅎㅎ

ChoJinmok commented on src/LoginPage/LoginForm.jsx:
```diff
@@ -10,10 +10,22 @@ const Container = styled.form({
+    lastName: {
+      value: lastNameValue,
+      invalidCheckMessage: lastNameInvalidCheckMessage,
+    },
```
`email.value`와 `emailValue`는 거의 차의가 없어서 이 부분이 조금 불필요해보이는 거 같습니당!!

ChoJinmok replied:
```javascript
fields: { email, password, firstName, lastName }
```

이런식으로 작성 후에 아래에서 `email.value`와 같이 사용하는 방법도 한번 고려해보세용!ㅎㅎ

xunxee replied:
나목 말 듣고 생각을 해보니, 제가 한 방법으로 
객체 구조 분해할당을 해서 얻을 수 있는 가독성의 이점이 
많이 떨어지는 것 같네용

위와 같은 경우에는, fields에서 말씀해주신 바와 같이
객체 구조분해할당을 해서 '마침표 표기법'으로 접근하는 것이
오히려 가독성이 더 좋을 것 같네요!
감사합니당🙏🏼

ChoJinmok commented on src/LoginPage/LoginForm.test.jsx:
```diff
@@ -16,19 +16,35 @@ describe('LoginForm', () => {
     isLogin,
     emailValue,
+    emailInvalidCheckMessage,
     passwordValue,
+    passwordInvalidCheckMessage,
     firstNameValue,
+    firstNameInvalidCheckMessage,
     lastNameValue,
+    lastNameInvalidCheckMessage,
     error,
   } = {}) {
     return render((
       <LoginForm
         isLogin={isLogin}
         fields={{
-          email: { value: emailValue },
-          password: { value: passwordValue },
-          firstName: { value: firstNameValue },
-          lastName: { value: lastNameValue },
+          email: {
+            value: emailValue,
+            invalidCheckMessage: emailInvalidCheckMessage,
+          },
+          password: {
+            value: passwordValue,
+            invalidCheckMessage: passwordInvalidCheckMessage,
+          },
+          firstName: {
+            value: firstNameValue,
+            invalidCheckMessage: firstNameInvalidCheckMessage,
+          },
+          lastName: {
+            value: lastNameValue,
+            invalidCheckMessage: lastNameInvalidCheckMessage,
+          },
           error,
         }}
         onChange={handleChange}
```
이 부분도 정리해줄 수 있을 거 같아요! fields 부분을 미리 풀어놓는 것이 오히려 코드를 길어지게 만드는 거 같습니당!

ChoJinmok replied:
```javascript
function renderLoginForm({ isLogin, fields } = {}) {
  return render((
    <LoginForm
      isLogin={isLogin}
      fields={fields}
      onChange={handleChange}
      onBlur={handleSignUpValid}
      onSubmit={handleSubmit}
    />
  ));
}

// ...

enderLoginForm({
  isLogin: true,
  fields: { error: 'not found' }
});

// ...

renderLoginForm({
  isLogin: false,
  fields: {
    [input.invalidCheckMessage]:
    `${inputs.inputName} is a required field.`,
  }
});
```

참고 하세용 ㅎㅎ

ChoJinmok commented on src/LoginPage/slice.js:
```diff
@@ -182,27 +182,38 @@ export function requestLogin() {
-    if (!value) return `${name} is a required field.`;
+export function checkSignUpValid({ name, value }) {
+  function validationCheckList({ inputValue }) {
+    if (!inputValue && name === 'lastName') {
+      return 'last name is a required field.';
+    }
+
+    if (!inputValue && name === 'firstName') {
+      return 'first name is a required field.';
+    }
+
+    if (!inputValue) {
+      return `${name} is a required field.`;
+    }
```
```javascript
if (!inputValue) {
  const nameFields = {
    lastName: 'last name',
    firstName: 'first name',
  }

  return `${nameFields[name] || name} is a required field.`
}
```

참고하세용 ㅎㅎ

xunxee replied:
호오... 좋은 코드는 끝이 없군요... 
반성합니다 ㅠㅠ

ChoJinmok commented on src/LoginPage/slice.test.js:
```diff
@@ -314,35 +298,67 @@ describe('actions', () => {
   });
 
   describe('checkSignUpValid', () => {
+    function makeLoginFields({
+      emailValue,
+      emailInvalidCheckMessage,
+      passwordValue,
+      passwordInvalidCheckMessage,
+      firstNameValue,
+      firstNameInvalidCheckMessage,
+      lastNameValue,
+      lastNameInvalidCheckMessage,
+    } = {}) {
+      return ({
+        loginFields: {
+          email: {
+            value: emailValue || '',
+            invalidCheckMessage:
+              emailInvalidCheckMessage || '',
+          },
+          password: {
+            value: passwordValue || '',
+            invalidCheckMessage:
+              passwordInvalidCheckMessage || '',
+          },
+          firstName: {
+            value: firstNameValue || '',
+            invalidCheckMessage:
+              firstNameInvalidCheckMessage || '',
+          },
+          lastName: {
+            value: lastNameValue || '',
+            invalidCheckMessage:
+              lastNameInvalidCheckMessage || '',
+          },
+        },
+      });
+    }
```
이 부분도 given으로 해결할 수 있을 거 같아용!!

ChoJinmok replied:
```javascript
function makeLoginFields({
  emailValue = '',
  emailInvalidCheckMessage = '',
  // ...
} = {}) {
  // ...
}
```

합연산자 사용안하고 이렇게 기본값 지정할 수 있는 점 참고하세용 ㅎㅎ

xunxee committed:
Add: slice, checkSignUpValid 함수의 비밀번호의 유효성을 검사하는 액션을 테스트 및 구현
- 정규표현식을 사용하여 회원가입시 비밀번호의 유효성을 검사

xunxee committed:
Refactor: LoginForm 컴포넌트의 props를 수정
- 기존의 fields의 모든 프로퍼티를 구조분해할당 받는 방식에서 마침표 표기법으로 접근하는 방법으로 코드의 가독성을 수정

xunxee committed:
Refactor: slice.js의 checkSignUpValid 함수의 로직을 수정
- if문 안에 객체를 선언하고 객체의 대괄호 표기법을 사용하여 값을 반환하는 방식으로 수정하여 가독성을 높이는 리팩터링을 진행

xunxee committed:
Refactor: slice.test.js의 checkSignUpValid 함수의 테스트를 리팩터링
- 기존의 loginFields의 객체를 반환하는 함수를 없애고, fixtures를 사용하여 mockStore를 호출

xunxee committed:
Refactor: slice.js의 checkSignUpValid 함수의 로직을 리팩터링
- 기존의 if문에서 validChecks 객체의 메서드를 사용하여 값을 반환하는 형식으로 코드를 리팩터링

xunxee committed:
Refactor: slice.test.js의 checkSignUpValid 함수의 테스트를 리팩터링
- profileInputs와 loginInputs를 만들어 배열을 순회하는 메서드인 forEach를 활용해 반복되는 테스트 코드를 리팩터링 작업

ChoJinmok commented on src/LoginPage/slice.test.js:
```diff
@@ -20,8 +20,8 @@ import { postLogin } from '../services/api';
 
-const middlewares = [thunk];
-const mockStore = configureStore(middlewares);
+const middleware = [thunk];
```
spelling checker가 에러를 잡는 것 같긴 한데 지금은 배열로 미들웨어들을 받는 거기 때문에 middleware's'가 맞는 거 같습니당!!

ChoJinmok commented on src/LoginPage/LoginForm.test.jsx:
```diff
@@ -31,19 +31,23 @@ describe('LoginForm', () => {
           email: {
             value: emailValue,
-            invalidCheckMessage: emailInvalidCheckMessage,
+            invalidCheckMessage:
+              emailInvalidCheckMessage,
           },
           password: {
             value: passwordValue,
-            invalidCheckMessage: passwordInvalidCheckMessage,
+            invalidCheckMessage:
+              passwordInvalidCheckMessage,
           },
           firstName: {
             value: firstNameValue,
-            invalidCheckMessage: firstNameInvalidCheckMessage,
+            invalidCheckMessage:
+              firstNameInvalidCheckMessage,
           },
           lastName: {
             value: lastNameValue,
-            invalidCheckMessage: lastNameInvalidCheckMessage,
+            invalidCheckMessage:
+              lastNameInvalidCheckMessage,
           },
           error,
         }}
```
slice에 있는 inputList를 fixtures에 추가

```javascript
// fixtures/inputList.js

const inputList = {
  lastName: 'Last Name은',
  firstName: 'First Name은',
  email: 'Email은',
  password: 'Password는',
};

export default inputList;
```

ChoJinmok replied:
```javascript
const fieldState = {
  value: '',
  invalidCheckMessage: '',
};

function renderLoginForm({
  isLogin = true,
  email = fieldState,
  password = fieldState,
  firstName = fieldState,
  lastName = fieldState,
  error = '',
} = {}) {
  return render((
    <LoginForm
      isLogin={isLogin}
      fields={{
        email,
        password,
        firstName,
        lastName,
        error,
      }}
      onChange={handleChange}
      onBlur={handleSignUpValid}
      onSubmit={handleSubmit}
    />
  ));
}

// ...

const { queryByText } = renderLoginForm({
  email: {
    value: 'tester',
    invalidCheckMessage: '',
  },
});

// ...

import INPUT_LIST from '../../fixtures/inputList';

const inputNames = [ lastName, firstName, email, password ];

it('renders invalid message', () => {
const { queryByText } = renderLoginForm({
  [inputName]: {
    value: '',
    invalidCheckMessage: `${INPUT_LIST[inputName]} 필수 입력란입니다.`
  }
});
```

ChoJinmok replied:
1. renderLoginForm의 매개변수에 기본값 지정하셔서 isLogin에 ture값 넣어주신 부분들 생략 가능할 듯 합니다!
2. 지금과 같이 매개변수의 기본값을 지정해놓지 않더라도 테스트는 통과할 수 있지만 인자를 넣어주지 않게되면 undefined가 들어가서 실제 상황과는 다른 상황이 만들어질 거 같습니다!
3. inputList의 경우 여러군데서 사용할 수 있을 거 같아서 fixtures에 추가했습니다! slice에서도 import 하셔서 사용해도 될 거 같습니다!

xunxee replied:
renderLoginForm 함수의 default 값을 지정해주는 것으로 코드의 가독성이 정말 많이 좋아지네요!
테스트를 통과하더라도 다음에 혹시 모를 확장성을 고려해서 default parameter를 지정하는 것이 맞는 것 같습니당
감사합니다🙏🏼

xunxee committed:
Refactor: LoginForm.test.js, renderLoginForm 함수의 로직을 리팩터링
- fixtures, inputList.js를 반복되는 코드를 개선
- renderLoginForm 함수의 parameter의 default 값을 선언하여 반복되는 코드를 리팩터링

xunxee committed:
Add: slice.js의 requestSignUp 함수의 액션을 테스트 및 구현

xunxee committed:
Refactor: slice.test.js의 에러를 수정
